flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java

```diff
@@ -0,0 +1,156 @@
+
+	Optional<CheckpointTriggerRequest> chooseRequestToExecute(CheckpointTriggerRequest newRequest, boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(false, () -> newRequest, isTriggering, lastCompletionMs);
+	}
+
+	Optional<CheckpointTriggerRequest> chooseQueuedRequestToExecute(boolean isTriggering, long lastCompletionMs) {
+		return chooseRequestToExecute(true, triggerRequestQueue::peek, isTriggering, lastCompletionMs);
+	}
+
+	/**
+	 * Choose the next {@link CheckpointTriggerRequest request} to execute based on the provided candidate and the
+	 * current state. Acquires a lock and may update the state.
+	 * @return request to execute, if any.
+	 */
+	private Optional<CheckpointTriggerRequest> chooseRequestToExecute(
+			boolean isEnqueued,
+			Supplier<CheckpointTriggerRequest> candidateSupplier,
+			boolean isTriggering,
+			long lastCompletionMs) {
+		return sync.apply(() -> {
+			final CheckpointTriggerRequest candidate = candidateSupplier.get();
+			if (candidate == null) {
+				return Optional.empty();
+			} else if (isTriggering) {
+				return onCantTriggerNow(candidate, isEnqueued);
+			} else if (pendingCheckpointsSizeSupplier.get() >= maxConcurrentCheckpointAttempts) {
+				return onCantCheckpointNow(candidate, TOO_MANY_CONCURRENT_CHECKPOINTS, () -> { /* not used */ }, isEnqueued);
+			} else if (nextTriggerDelayMillis(lastCompletionMs) > 0) {
+				return onCantCheckpointNow(candidate, MINIMUM_TIME_BETWEEN_CHECKPOINTS, () -> rescheduleTrigger.accept(nextTriggerDelayMillis(lastCompletionMs)), isEnqueued);
+			} else if (candidate != triggerRequestQueue.peek()) {
+				triggerRequestQueue.offer(candidate);
+				return Optional.of(triggerRequestQueue.poll());
+			} else {
+				return Optional.of(isEnqueued ? triggerRequestQueue.poll() : candidate);
+			}
+		});
+	}
+
+	private Optional<CheckpointTriggerRequest> onCantTriggerNow(CheckpointTriggerRequest candidate, boolean isEnqueued) {
+		offerIfNeeded(candidate, isEnqueued);
+		return Optional.empty();
+	}
+
+	private Optional<CheckpointTriggerRequest> onCantCheckpointNow(
+			CheckpointTriggerRequest candidate,
+			CheckpointFailureReason dropReason,
+			Runnable postDrop,
+			boolean isEnqueued) {
+		if (candidate.props.forceCheckpoint()) {
+			return Optional.of(isEnqueued ? triggerRequestQueue.poll() : candidate);
+		} else if (candidate.isPeriodic) {
+			if (isEnqueued) {
+				triggerRequestQueue.poll();
+			}
+			candidate.onCompletionPromise.completeExceptionally(new CheckpointException(dropReason));
+			postDrop.run();
+		}
+		offerIfNeeded(candidate, isEnqueued);
+		return Optional.empty();
+	}
+
+	private void offerIfNeeded(CheckpointTriggerRequest candidate, boolean isEnqueued) {
+		if (!isEnqueued) {
+			if (triggerRequestQueue.size() < MAX_QUEUED_REQUESTS) {
+				triggerRequestQueue.offer(candidate);
+			} else {
+				candidate.onCompletionPromise.completeExceptionally(new CheckpointException(CheckpointFailureReason.TOO_MANY_CHECKPOINT_REQUESTS));
```

Comment:
I introduced this check only to ease debugging in case of bugs (or probably misconfiguration). 
This shouldn't happen normally so I don't think testing it adds enough value.

Reply:
I would still add a test for that. It's a valid feature and handling the corner/error cases is an important thing.

Reply:
Added a test.